yb-voyager/src/callhome/diagnostics.go

```diff
@@ -0,0 +1,137 @@
+	TargetDBCores         int       `json:"target_db_cores"`         //unknown for now
+	SourceCloudDBType     string    `json:"source_cloud_type"`       //unknown for now
+}
+
+// Fill in primary-key based fields, if needed
+func InitJSON(exportdir string) {
+
+	jsonFilePath = filepath.Join(exportdir, "metainfo", "diagnostics.json")
+	file, err := os.OpenFile(jsonFilePath, os.O_RDWR|os.O_CREATE, 0644)
+	file.Close()
+	if err != nil {
+		utils.ErrExit("Error while creating/opening diagnostics.json file: %v", err)
+	}
+	jsonBuf, err = os.ReadFile(jsonFilePath)
+	if err != nil {
+		utils.ErrExit("Error while reading diagnostics.json file: %v", err)
+	}
+
+	if len(jsonBuf) != 0 {
+		err = json.Unmarshal(jsonBuf, &Payload)
+		if err != nil {
+			utils.ErrExit("Invalid diagnostics.json file: %v", err)
+		}
+	}
+
+	if Payload.MigrationUuid == uuid.Nil {
+		Payload.MigrationUuid, err = uuid.NewUUID()
+		Payload.StartTime = time.Now().Format("2006-01-02 15:04:05")
+		if err != nil {
+			utils.ErrExit("Error while generating new UUID for diagnostics.json: %v", err)
+		}
+	}
+
+}
+
+// Getter method for updating payload
+func GetPayload() *payload {
+	return &Payload
+}
+
+// Send http request to flask servers after saving locally
+func PackAndSendPayload(exportdir string) {
+	if !SendDiagnostics {
+	} else {
+		//Pack locally
+		jsonBuf, _ = json.Marshal(Payload)
+		os.WriteFile(jsonFilePath, jsonBuf, 0644)
+
+		//Send request
+		Payload.LastUpdatedTime = time.Now().Format("2006-01-02 15:04:05")
+		postBody, _ := json.Marshal(Payload)
+		requestBody := bytes.NewBuffer(postBody)
+
+		log.Infof("Payload being sent for diagnostic usage: %s\n", string(postBody))
+		resp, err := http.Post("http://10.150.5.149:5000/", "application/json", requestBody)
+
+		if err != nil {
+			utils.ErrExit("Error while sending diagnostic data: %v", err)
+		}
+
+		defer resp.Body.Close()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			utils.ErrExit("Error while reading HTTP response from call-home server: %v", err)
+		}
+		log.Infof("HTTP response after sending diagnostic.json: %s\n", string(body))
+	}
+}
+
+// Find the largest and total data sizes, and upload to diagnostics json
+func UpdateDataSize(exportdir string) {
+	datadirfiles := filepath.Join(exportdir, "data", "*_data*")
+
+	files, err := filepath.Glob(datadirfiles)
+	if err != nil {
+		utils.ErrExit("Error while matching files in data dir for diagnostics: %v", err)
+	}
+	var totalSize int64
+	var maxFileSize int64
+	for _, file := range files {
+		fileInfo, err := os.Stat(file)
+		if err != nil {
+			utils.ErrExit("Error while querying files for size: %v", err)
+		}
+		if maxFileSize < fileInfo.Size() {
+			maxFileSize = fileInfo.Size()
+		}
+		totalSize += fileInfo.Size()
+	}
+
+	Payload.TotalSize = totalSize
+	Payload.LargestTableSize = maxFileSize
+
+	PackAndSendPayload(exportdir)
```

Comment:
This should be moved out of this function and must be called from one level up.

Reply:
Will do this, just noticed it will make the code readability more consistent 👍🏽